Classes/common/CDTIncrementalStore.m

```diff
@@ -0,0 +1,2798 @@
+ *  1. I can continue from this simply by typing:
+ *  ```
+ *  strap register write pc `$pc+2`
+ *  ```
+ *  > Different architectures will use different addend values
+ *  2. I don't need to "Add Exception Breakpoint"
+ *  3. I don't need to hunt down which exception a test is using in an
+ *  expected way
+ *
+ *  *Why is it a macro?*
+ *  I want to stop *at* the `oops` line in the code and not have to "pop up"
+ *  the stack if `oops` was not inlines due to optimization issues.
+ *
+ *  @param fmt A format string
+ *  @param ... A comma-separated list of arguments to substitute into format.
+ */
+#define oops(fmt, ...)                                                     \
+    do {                                                                   \
+        NSLog(@"%s:%u OOPS: %s", __FILE__, __LINE__, __PRETTY_FUNCTION__); \
+        NSLog(fmt, ##__VA_ARGS__);                                         \
+        fflush(stderr);                                                    \
+        __builtin_trap();                                                  \
+    } while (NO);
+
+static NSString *stringFromData(NSData *data)
+{
+    NSMutableString *s = [NSMutableString string];
+    const unsigned char *d = (const unsigned char *)[data bytes];
+    size_t sz = [data length];
+
+    for (size_t i = 0; i < sz; i++) {
+        [s appendString:[NSString stringWithFormat:@"%02x", d[i]]];
+    }
+    return [NSString stringWithString:s];
+}
+
+static NSData *dataFromString(NSString *str)
+{
+    char buf[3] = {0};
+
+    size_t sz = [str length];
+
+    if (sz % 2) {
+        oops(@"must be even number of characters (%zd): %@", sz, str);
+    }
+
+    unsigned char *bytes = malloc(sz / 2);
+    unsigned char *bp = bytes;
+    for (size_t i = 0; i < sz; i += 2) {
+        buf[0] = [str characterAtIndex:i];
+        buf[1] = [str characterAtIndex:i + 1];
+        char *chk = NULL;
+        *bp = strtol(buf, &chk, 16);
+        if (chk != buf + 2) {
+            oops(@"bad character around %zd: %@", i, str);
+        }
+        ++bp;
+    }
+
+    return [NSData dataWithBytesNoCopy:bytes length:sz / 2 freeWhenDone:YES];
+}
+
+@interface CDTISProperty : NSObject
+// Information about the object
+@property (nonatomic) BOOL isRelationship;
+@property (strong, nonatomic) NSString *typeName;
+@property (nonatomic) NSInteger typeCode;
+@property (strong, nonatomic) NSString *xform;
+@property (strong, nonatomic) NSString *destination;
+@property (strong, nonatomic) NSData *versionHash;
+
+- (CDTISProperty *)initWithAttribute:(NSAttributeDescription *)attribute;
+- (CDTISProperty *)initWithRelationship:(NSRelationshipDescription *)relationship;
+- (NSDictionary *)dictionary;
+@end
+
+@implementation CDTISProperty
+
+/**
+ * Defines tha attribute meta data that is stored in the object.
+ *
+ *  @param att
+ *
+ *  @return initialized object
+ */
+- (CDTISProperty *)initWithAttribute:(NSAttributeDescription *)attribute
+{
+    self.isRelationship = NO;
```

Comment:
Another style issue with initializers: I think it is best to avoid using 'self.' when referencing properties in the initializer, since this requires that all getters and setters behave properly even on partially initialized objects.  It is better to set the underlying instance variables directly in the initializer to avoid this dependency.